### DISCUSSION TITLE
Support --no-optional at the package.json level

### DISCUSSION THREAD

**Discussion by dominic-p** [npm/feedback#395] [General]
This is basically a repost of [this old GitHub issue](https://github.com/npm/npm/issues/14185) which wasn't really resolved as far as I could tell.

It seems to be a common issue that we depend on some package that defines optional dependencies that we don't want installed. Running `npm install --no-optional` handles this nicely, but it's a pain to write every time, and it's not really fine grained (for some packages we actually do want optional dependencies installed).

It would be nice if there was a package.json level way to say that "I don't want optional dependencies installed for this dependency." Some of the ideas in the old thread were breaking changes that may be hard to implement. My humble suggestion would be to add a new field (`pkg.noOptional`) that could look something like this:

```
{
    "name": "my-modules",
    "dependencies": {
        "someone-elses": "^2.0.3",
        "has-bad-optionals": "^1.0.2",
        "mixed-bag": "^1.2.1"
    },
    "noOptional": {
        "has-bad-optionals": "*",
        "mixed-bag": ["bad-optional"]
    }
}
```

**Comment by MylesBorins** [comment]
@dominic-p you can set

```
optional = false
```

in a .npmrc file of a project an avoid optional dependencies to be installed at all for that project.

That said it looks like you are looking for a slightly more robust and potentially per dependency way to manage this. 

If you want to suggest new CLI features the best next step is to open a issue over at https://github.com/npm/rfcs

**Reply by dominic-p** [reply]
Thanks for the suggestion @MylesBorins I appreciate it. That definitely works as a band-aid for now, but as you mentioned, it would be nice to have a little more fine grained control. My issue for example is with a dependency of a library that is shared by multiple projects. I would love to suppress the bad optional dependencies at the library level instead of having to do it for each project that uses the library.

It seems like there was a lot of interest in the old thread, but maybe not so much anymore. Maybe the community found another solutions to this.
